### PR TITLE
[FEATURE] Table legend for time series chart panel

### PR DESCRIPTION
**Pull request by juliepagano**
The time series chart panel now supports the option to select a legend "mode" that is either a list (previous legend UI) or a table (new option).

# Notes for reviewers

- I added a dashboard that exercises the various modes at http://localhost:3000/projects/testing/dashboards/TimeSeriesLegends?start=5m if helpful for testing these changes. There's also a bunch in storybook.
- I'm intentionally leaving "list" mode alone at this time to be minimally disruptive until we add more sizes and other functionality.

# Checklist

- [x] Pull request has a descriptive title and context useful to a reviewer.
- [x] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `IGNORE`.
- [x] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

## UI Changes

- [x] Changes that impact the UI include screenshots and/or screencasts of the relevant changes.
- [x] Code follows the [UI guidelines](https://github.com/perses/perses/blob/main/ui/ui-guidelines.md).
- [x] Visual tests are stable and unlikely to be flaky. See [Storybook](https://github.com/perses/perses/tree/main/ui/storybook#visual-tests) and [e2e](https://github.com/perses/perses/tree/main/ui/e2e#visual-tests) docs for more details. Common issues include:
  - Is the data inconsistent? You need to mock API requests.
  - Does the time change? You need to use consistent time values or mock time utilities.
  - Does it have loading states? You need to wait for loading to complete.

# Screenshots

image
image

https://github.com/perses/perses/assets/396962/d65628ee-7945-4db3-bdc5-894d15640def

### CODE REVIEW

File: ui/components/src/ContentWithLegend/model/content-with-legend-model.test.ts
Comment (juliepagano) [inline ui/components/src/ContentWithLegend/model/content-with-legend-model.test.ts:39]: I put a loop around these tests to check both list and table, which made the diff a little funky, but the tests haven't changed much since the previous PR.

File: ui/components/src/ContentWithLegend/model/content-with-legend-model.ts
Comment (juliepagano) [inline ui/components/src/ContentWithLegend/model/content-with-legend-model.ts:149]: TODO: get the row height from the legend utils instead of hardcoding it here.

File: ui/components/src/ContentWithLegend/model/content-with-legend-model.ts
Comment (juliepagano) [inline ui/components/src/ContentWithLegend/model/content-with-legend-model.ts:137]: This avoids this being really disruptive to current list users who were used to the previous size. We default the table to what will be "medium" when we add the new sizes.

File: ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanelComponent.stories.tsx
Comment (juliepagano) [inline ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanelComponent.stories.tsx:103]: Changed this to a mock that shows a bunch of series, which is better for showing off the legends.

File: schemas/panels/time-series/mig.cuepart
Comment (juliepagano) [inline schemas/panels/time-series/mig.cuepart:9]: This is my first time changing these mig.cuepart files, and I wasn't 100% sure how to test that they work. Any pointers?

File: schemas/panels/time-series/time-series.cue
Comment (juliepagano) [inline schemas/panels/time-series/time-series.cue:22]: Making this optional, so the addition of the table mode isn't a breaking change for consumers.

File: schemas/panels/time-series/mig.cuepart
Comment (juliepagano) [reply schemas/panels/time-series/mig.cuepart:9]: Aha, I think I found some failing go tests that answer my question here.

File: internal/api/shared/migrate/testdata/simple_grafana_dashboard.json
Comment (juliepagano) [inline internal/api/shared/migrate/testdata/simple_grafana_dashboard.json:335]: Swapping one of these to "table" so we exercise both list and table modes.

File: ui/components/src/ContentWithLegend/ContentWithLegend.stories.tsx
Comment (juliepagano) [inline ui/components/src/ContentWithLegend/ContentWithLegend.stories.tsx:156]: Added positions so you can see the combo of all modes and all positions.

File: ui/components/src/ContentWithLegend/model/content-with-legend-model.ts
Comment (juliepagano) [inline ui/components/src/ContentWithLegend/model/content-with-legend-model.ts:141]: Logic for the "list" mode is largely unchanged besides the vars being declared above.

File: dev/data/dashboard.json
Comment (juliepagano) [inline dev/data/dashboard.json:3061]: I didn't add any e2e tests for this because we get pretty good coverage from jest tests and visual tests from storybook. However, I added a test dashboard for ease of debugging and playing around with legends in the app context.

**Comment by eunicorn** [comment]
Is this expected? Looks a bit weird with the graph being so squished

image

**Comment by juliepagano** [comment]
> Is this expected? Looks a bit weird with the graph being so squished
> 
> image

It is currently. The expectation is that at the "medium" size (what we're starting with), the bottom-positioned table is 5 rows tall. We'll handle adding in a "small" size in a follow-up. As part of that, we should also probably add some special handling to make it smaller if there are less than 5 rows (there's been a TODO comment to do something similar for the list legend for ages). The right-positioned table looks a little nicer at this point.

There's also some open questions about how big the panel rendering should be in the settings, so it can better reflect what you might see when you put it onto the dashboard. Some conversations about that prompted @cndonovan to start on https://github.com/perses/perses/pull/1208. 

TLDR: we should do more here in follow-ups, but I'm hoping this PR can be incremental progress for delivering the table legend. Let me know if you think it's critical to resolve some of these issues in this PR instead of in follow-ups.

**Comment by sjcobb** [comment]
The y_axis.label bug you pointed out before where it isn't centered when the legend is in Bottom position becomes a bit more urgent to fix now, on one of my test dashboards changing legend mode resulted in:
image

**Review by sjcobb — APPROVED** [review]
LGTM! I think the [screenshot](https://github.com/perses/perses/pull/1197#issuecomment-1573023916) I added can be it's own ticket and fixed separately since it's more a problem with how I implemented YAxisLabel [positioning](https://github.com/perses/perses/blob/70bcb3856bc4b53bc3a37abb31d23fcb64ec5aad/ui/components/src/YAxisLabel.tsx#L28) than anything to do with the legend.

Planning to do release today so hopefully this can make it in! 🚀

**Review by eunicorn — APPROVED** [review]

**Comment by juliepagano** [comment]
> LGTM! I think the [screenshot](https://github.com/perses/perses/pull/1197#issuecomment-1573023916) I added can be it's own ticket and fixed separately since it's more a problem with how I implemented YAxisLabel [positioning](https://github.com/perses/perses/blob/70bcb3856bc4b53bc3a37abb31d23fcb64ec5aad/ui/components/src/YAxisLabel.tsx#L28) than anything to do with the legend.
> 
> Planning to do release today so hopefully this can make it in! 🚀

Do you want to hold off on doing a release until early Monday since I think a bunch of the maintainers are out in the afternoon? I think there will be a handful of breaking changes in this one, so it'd be nice to have folks around to give it a close eye.

**Comment by juliepagano** [comment]
> The y_axis.label bug you pointed out before where it isn't centered when the legend is in Bottom position becomes a bit more urgent to fix now, on one of my test dashboards changing legend mode resulted in: 
> 
> Also, possibly some z-index adjustments will be needed so that the y_axis label shows on top or gets truncated too in these cases.

Good call! I think we're going to try to give that bug to one of our new teammates next week, since it'll be a nice intro-to-perses task to pick up.

**Pull request merged**
